guidelines.rst

```diff
@@ -154,6 +154,134 @@ The more stable the tier is, the stricter the rules for it are:
+
+    - https://github.com/capi-workgroup/api-evolution/issues/34
+
+
+.. XXX add a PEP number prefix to the anchor:
+
+.. _naming:
+
+Naming
+======
+
+[TODO: PEP 7 should be updated to link here once this PEP goes live.]
+
+All public names must be prefixed with ``Py``.
+
+Names that users should not use directly, but need to be visible to the
+compiler/linker, should be prefixed with ``_Py``.
+(Such names are not considered public API, that is, they should not appear in
+third-party source code.)
+
+This applies to all names in a global namespace: functions, macros, variables,
+typedefs, structs, enums, etc.; not to parameters or struct fields.
+
+The ``Py_`` prefix is reserved for global service routines like
+``Py_FatalError``; specific groups of APIs use a longer prefix,
+for example ``PyUnicode_`` for string functions.
+Use an existing prefix when applicable. If you want to add a new prefix,
+contact the C API Working Group.
+
+The ``Py_`` prefix is in mixed case, even in macro names.
+For example: ``PyUnicode_AS_STRING``.
+(Several existing macros use the upper-case ``PY``; if you need this prefix
+for consistency, please get approval from the C API Working Group.)
+
+Unstable API is prefixed with ``PyUnstable_`` instead of ``Py``,
+for example ``PyUnstable_Long_IsCompact`` or (hypothetically)
+``PyUnstable_String_GET_SIZE``.
+
+.. note::
+
+    For background and discussions, see:
+
+    - https://github.com/capi-workgroup/api-evolution/issues/21
+
+
+Do not reuse names
+------------------
+
+If an API's behavior changes in a backwards-incompatible way,
+for example when a function's signature changes,
+add new API with a new name.
+You can deprecate and remove the old version following Python's
+:pep:`backwards compatibility policy <387>`.
+
+After API has been removed, do not reuse the old name,
+since existing documentation and tutorials will continue to refer to the
+old behavior.
+
+
+Language support
+================
+
+C standard and dialect
+----------------------
+
+[TODO: PEP 7 should be updated to link here once this PEP goes live.]
+
+Public C API must be compatible with:
+
+- C11, with optional features needed by CPython:
+
+  - IEEE 754 floating point
+  - Atomics (``!__STDC_NO_ATOMICS__``, or MSVC)
+
+- C99
+- C89 with several select C99 features:
+
+  - ``<stdint.h>`` and ``<inttypes.h>``
+  - ``static inline`` functions
+  - designated initializers
+  - intermingled declarations
+  - line comments (``//``)
+
+- C++03
+
+It is OK to use other features -- compiler-specific ones,
+optional standard ones, or platform-specific ones -- if the behavior of
+correct user code is the same as with a standard compiler.
+For example, compiler-specific code is often used to improve performance
+or compiler diagnostics.
+
+It is also OK to use these other features for platform-specific API,
+which needs to be documented as such and have a feature test macro
+(for example, ``PY_HAVE_THREAD_NATIVE_ID``).
```

Comment:
PY_ prefix is a bad example :-) Maybe use Py_HAVE_C_COMPLEX example.

Reply:
IMO, undocumented API is worse than an older naming style.

Reply:
> IMO, undocumented API is a worse example than one with an older naming style.

`PY_HAVE_THREAD_NATIVE_ID` is not documented.

Reply:
Silly me.